contracts/mainnet/connectors/morpho-blue/main.sol

```diff
@@ -0,0 +1,918 @@
+			TokenInterface(wethAddr),
+			_amt
+		);
+
+		setUint(_setId, _amt);
+
+		_eventName = "LogWithdrawCollateral((address,address,address,address,unit256),uint256,uint256,uint256)";
+		_eventParam = abi.encode(_marketParams, _amt, _getId, _setId);
+	}
+
+	/**
+	 * @notice Handles the withdrawal of collateral by a user from a specific market of a specific amount.
+	 * @dev The market to withdraw assets from. (For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param _marketParams The market to withdraw assets from.
+	 * @param _assets The amount of assets to withdraw. (For max: `uint256(-1)`)
+	 * @param _onBehalf The address that already deposited position.
+	 * @param _getId ID to retrieve amt.
+	 * @param _setId ID stores the amount of tokens deposited.
+	 */
+	function withdrawCollateralOnBehalf(
+		MarketParams memory _marketParams,
+		uint256 _assets,
+		address _onBehalf,
+		address _receiver,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _assets);
+
+		_marketParams = updateTokenAddresses(_marketParams);
+
+		// If amount is max, fetch collateral value from Morpho's contract
+		if (_amt == type(uint256).max) {
+			bytes32 _id = id(_marketParams);
+			Position memory _pos = MORPHO_BLUE.position(_id, _onBehalf);
+			_amt = _pos.collateral;
+		}
+
+		MORPHO_BLUE.withdrawCollateral(
+			_marketParams,
+			_amt,
+			_onBehalf,
+			_receiver
+		);
+
+		if (_receiver == address(this))
+			convertWethToEth(
+				_marketParams.collateralToken == ethAddr,
+				TokenInterface(wethAddr),
+				_amt
+			);
+
+		setUint(_setId, _amt);
+
+		_eventName = "LogWithdrawCollateralOnBehalf((address,address,address,address,unit256),uint256,address,address,uint256,uint256)";
+		_eventParam = abi.encode(
+			_marketParams,
+			_amt,
+			_onBehalf,
+			_receiver,
+			_getId,
+			_setId
+		);
+	}
+
+	/**
+	 * @notice Handles the withdrawal of supply.
+	 * @dev  The market to withdraw assets from. (For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param _marketParams The market to withdraw assets from.
+	 * @param _assets The amount of assets to withdraw. (For max: `uint256(-1)`)
+	 * @param _getId ID to retrieve amt.
+	 * @param _setId ID stores the amount of tokens deposited.
+	 */
+	function withdraw(
+		MarketParams memory _marketParams,
+		uint256 _assets,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _assets);
+
+		_marketParams = updateTokenAddresses(_marketParams);
+
+		if (_amt == type(uint256).max) {
+			bytes32 _id = id(_marketParams);
+			Position memory _pos = MORPHO_BLUE.position(_id, address(this));
+			uint256 _shares = _pos.supplyShares;
+
+			_amt = _toAssetsUp(
+				_shares,
+				MORPHO_BLUE.market(_id).totalSupplyAssets,
+				MORPHO_BLUE.market(_id).totalSupplyShares
+			);
+		}
```

Comment:
when withdrawing max it's advised to use shares instead to make sure no dust is left on the contract

Reply:
Got it!
Updated the logics to use shares in case of max amounts. [Here](https://github.com/Instadapp/dsa-connectors-2.0/pull/1/files#diff-920626a6b7a6adcbdc1062029932b5a6d9b5ee7f2a461452bc13e62a5b7274b1R437-R453)